### PR TITLE
Merge `vellottie`

### PR DESCRIPTION
**Pull request by simbleau**
This PR will merge in changes from vellottie - 
It will also bring velato to a state where it can almost be published.

#14 will bring a followup to continue to prepare for release.

### CODE REVIEW

File: examples/with_winit/README.md
Comment (DJMcNab) [inline examples/with_winit/README.md:3]: I don't think the (one?) built in file is public domain in the legal sense

File: examples/with_winit/src/hot_reload.rs
Comment (DJMcNab) [inline examples/with_winit/src/hot_reload.rs:1]: This file can go

File: examples/with_winit/src/lib.rs
Comment (DJMcNab) [inline examples/with_winit/src/lib.rs:1]: I've only skimmed this file, but nothing jumps out

File: rustfmt.toml
Comment (DJMcNab) [inline rustfmt.toml:1]: This is the default value

File: rustfmt.toml
Comment (DJMcNab) [inline rustfmt.toml:2]: Nice, I didn't know you could set these on stable. These seem like a simple improvement

I don't really like this being at the root, but that's also not avoidable

File: src/parser/schema/animated_properties/animated_property.rs
Comment (DJMcNab) [inline src/parser/schema/animated_properties/animated_property.rs:1]: I'd generally expect these to all be within one file.

That is, I'd prefer for each folder here to be one file, given that the size of each of these is only 1-2 structs.

File: src/import/mod.rs
Comment (DJMcNab) [inline src/import/mod.rs:1]: I'd expect this to be at `src/import.rs`. That's the convention used in the `examples` dir, for example

File: src/import/util.rs
Comment (DJMcNab) [inline src/import/util.rs:7]: Should this be `pub(crate)`?

File: src/lib.rs
Comment (DJMcNab) [inline src/lib.rs:1]: We probably should have some usage docs here or in the readme

File: .github/workflows/ci.yml
Comment (DJMcNab) [inline .github/workflows/ci.yml:106]:
```suggestion
```

**Review by DJMcNab — COMMENTED** [review]
Lots of small things which need addressing, but overall this works well for me locally

I feel like there is a better way to present the downloads (e.g. in a clickable grid showing them all running at the same time?)

But that can be a follow-on, if you agree that would be good.

**Comment by DJMcNab** [comment]
I've also seen that the downloads aren't included in the default runner by default.

I'm not sure the best way to handle that is - we should probably not panic, but yes print the download message?

File: .github/workflows/pages-release.yml
Comment (simbleau) [reply .github/workflows/pages-release.yml:6]: Actually, I'd rather have it show the latest release.

Comment (simbleau) [reply .github/workflows/pages-release.yml]: Just updated to match `Vello`'s setting.

File: AUTHORS
Comment (simbleau) [reply AUTHORS:1]: It was copy/pasted

File: src/parser/schema/animated_properties/animated_property.rs
Comment (simbleau) [reply src/parser/schema/animated_properties/animated_property.rs:1]: Right now I want to defer schema changes. Let's ignore that.

The schema for lottie has been recently changing due to the LAC (Lottie Animation Committee) and you can already see some significant organization changes between these two files:

Original schema: https://lottiefiles.github.io/lottie-docs/schema/lottie.schema.json
LAC schema: https://lottie.github.io/lottie-spec/lottie.schema.json 

Pretty soon I want to re-organize this anyways, to match schema parity. I can incorporate this feedback when we get there.

File: src/import/util.rs
Comment (simbleau) [reply src/import/util.rs:7]: Not needed, because `util` is not public.

File: AUTHORS
Comment (simbleau) [reply AUTHORS]: fixed

File: AUTHORS
Comment (simbleau) [reply AUTHORS]: no, fixed

File: README.md
Comment (simbleau) [reply README.md:11]: fixed

File: README.md
Comment (simbleau) [reply README.md]: fixed

File: README.md
Comment (simbleau) [reply README.md]: fixed

File: README.md
Comment (simbleau) [reply README.md]: fixed

File: examples/assets/Tiger.json
Comment (simbleau) [reply examples/assets/Tiger.json:1]: fixed - included the license now

File: examples/scenes/src/download/default_downloads.rs
Comment (simbleau) [reply examples/scenes/src/download/default_downloads.rs:11]: Feedback accepted - There are no non-noto items here

File: examples/scenes/src/lottie.rs
Comment (simbleau) [reply examples/scenes/src/lottie.rs:101]: It is fast enough

File: README.md
Comment (simbleau) [reply README.md:25]: fixed

File: README.md
Comment (simbleau) [reply README.md:70]: fixed (included)

File: examples/scenes/src/download.rs
Comment (simbleau) [reply examples/scenes/src/download.rs:72]: Do we really *need* to do this? 

Personally I think the with_winit demo should be abstracted into something we can import here, and vello_svg - And keep that work TODO when we get around to that.

File: src/import/mod.rs
Comment (simbleau) [reply src/import/mod.rs:1]: This can be a followup task addressed in #14

File: src/parser/schema/animated_properties/animated_property.rs
Comment (simbleau) [reply src/parser/schema/animated_properties/animated_property.rs:1]: This can be a followup task addressed in #14

File: src/lib.rs
Comment (simbleau) [reply src/lib.rs:1]: This can be a followup task addressed in #14

File: examples/run_wasm/src/main.rs
Comment (simbleau) [reply examples/run_wasm/src/main.rs:22]: fixed

File: examples/scenes/src/download/default_downloads.rs
Comment (simbleau) [reply examples/scenes/src/download/default_downloads.rs:11]: This can be a followup task addressed in #14

File: examples/with_winit/Cargo.toml
Comment (simbleau) [reply examples/with_winit/Cargo.toml:26]: This can be a followup task addressed in https://github.com/linebender/velato/issues/14

File: examples/with_winit/README.md
Comment (simbleau) [reply examples/with_winit/README.md]: fixed

File: examples/scenes/src/download.rs
Comment (DJMcNab) [reply examples/scenes/src/download.rs:72]: The point of printing the sizes is in case downloads are going to take up too much disk space/network usage - if e.g. you're on a limited internet connection

Similarly, the point of printing the licenses is so that people know what they're agreeing to

Checking that 350 items have the sameish size (as there is surely some variation), then counting them and multiplying. Then manually checking that each license is the same is pointless busiwork.

My concern is not about how long the downloads will take

File: src/import/util.rs
Comment (DJMcNab) [reply src/import/util.rs:7]: You're right, we should add the unreachable_pub lint

File: .gitignore
Comment (simbleau) [reply .gitignore:13]: Why does a newline at the end matter?

File: Cargo.toml
Comment (simbleau) [reply Cargo.toml:22]: Will do - Vello uses this wording:

```
# NOTE: Make sure to keep this in sync with the version badge in README.md
wgpu = { version = "0.19.3" }
```

File: README.md
Comment (DJMcNab) [reply README.md:18]: I think we can leave it like this in this PR, though

File: Cargo.toml
Comment (xStrom) [inline Cargo.toml:23]: Let's specify the full version because otherwise `cargo upgrade` will ignore patch updates.

```suggestion
vello = "0.1.0"
```

File: Cargo.toml
Comment (xStrom) [inline Cargo.toml:36]: Let's specify the full version because otherwise `cargo upgrade` will ignore patch updates. Also, some of these are really old, e.g. serde 1.0 was released in 2017. Someone might try to actually use it via `-Z minimal-versions` and we don't test it.

I'll note that I haven't actually tested this combination locally. Hopefully no deep dependency has a pinned sub dependency which prevents using the latest versions. Our CI will catch that in this PR if that's the case though.

```suggestion
keyframe = "1.1.1"
once_cell = "1.19.0"

# For the parser
serde = { version = "1.0.197", features = ["derive"] }
serde_json = "1.0.114"
serde_repr = "0.1.18"

[target.'cfg(target_arch = "wasm32")'.dev-dependencies]
wasm-bindgen-test = "0.3.42"
```

File: Cargo.toml
Comment (simbleau) [reply Cargo.toml:23]: This is interesting to hear... I thought by putting 0.1, anyone who uses Velato as a library can use any patch version, which gives maximum flexibility. Is this not what we want?

Comment (xStrom) [reply Cargo.toml:23]: So, for someone using this library, there is absolutely zero difference between `0.1` and `0.1.0`. The difference is in relation to `cargo upgrade` (via [`cargo-edit`](https://crates.io/crates/cargo-edit)) which only modifies the numbers that exist. It's about maintenance for us, not about people actually using the library.

Comment (simbleau) [reply Cargo.toml:23]: Is there ever a reason to use 0.1 or serde = "1"?

Comment (xStrom) [reply Cargo.toml:23]: Reasons to not write the patch version (`0.1`) can include wanting to write fewer characters, easier reading, and signaling that you don't care about patch releases.

Serde `1` is wrong more often than not, because it makes a claim that `1.0.0` is sufficient which is often not the case with libraries as old as Serde. See [this thread](https://users.rust-lang.org/t/psa-please-specify-precise-dependency-versions-in-cargo-toml/71277) for one explanation.

File: Cargo.toml
Comment (simbleau) [reply Cargo.toml]: fixed

File: .gitignore
Comment (simbleau) [reply .gitignore]: fixd

File: examples/assets/LICENSE
Comment (simbleau) [reply examples/assets/LICENSE:1]: fixed

File: examples/scenes/src/download/default_downloads.rs
Comment (simbleau) [reply examples/scenes/src/download/default_downloads.rs]: fixed with a macro

File: examples/scenes/src/download.rs
Comment (simbleau) [reply examples/scenes/src/download.rs:72]: I'll add it as a followup task in #14

File: Cargo.toml
Comment (DJMcNab) [inline Cargo.toml:10]:
```suggestion
publish = false
```
This is likely to be publishable as-is, but it's better to make sure we have consensus before doing that (e.g. an opportunity to review the final state of the readme, etc.)

Comment (simbleau) [reply Cargo.toml:10]: This doesn't stop someone from publishing, e.g. "cargo publish --allow-dirty" - I'd rather just have a conversation before publishing but I'll keep the comments on this open.

Comment (DJMcNab) [reply Cargo.toml:10]: This suggestion is not about stopping someone malicious publishing - we can't do that

It's specifically about making sure there's consensus for an initial release. E.g. I'd want Chad to sign off on making a release, whereas this PR doesn't require their input.

Comment (simbleau) [reply Cargo.toml:10]: @dfrg

Comment (DJMcNab) [reply Cargo.toml:10]: Whether that consensus is gained on GitHub or on Zulip, it's easier to track and record that on GitHub with the standardised approval button

**Comment by dfrg** [comment]
This is rather large and I’m currently otherwise occupied, but at a glance, I don’t see any major blockers and I’ll review this by the of the week.

Thanks @simbleau for taking this on!

File: Cargo.toml
Comment (dfrg) [reply Cargo.toml:10]: I’m mostly ambivalent on this but adding a tiny safeguard makes sense to me so +1 to Daniel’s suggestion.

**Comment by dfrg** [comment]
I haven’t forgotten about this. I’ll be taking a look shortly.

**Review by dfrg — COMMENTED** [review]
Thanks Spencer! Overall this looks like a set of solid improvements. I added some notes on things I'd like changed before merging, suggestions for future work and a few random comments.

I'm happy to land this after the requested changes are addressed.

File: src/import/converters.rs
Comment (dfrg) [inline src/import/converters.rs:707]: Just as an example, replacing `Number` with `f64` reduces this to
```rust
Static(components) => Value::Fixed(f(&components)),
```

File: src/runtime/render.rs
Comment (simbleau) [reply src/runtime/render.rs:83]: Yes, I found there to be significant benefit to choosing a frame to render instead of seconds.

I made this change when I implemented the dotLottie spec - which is built for interactivity. It was much easier to control a playhead (frame) rather than seconds, because of things like slow motion, bounce play mode, etc.

Also, in the seconds implementation, it would just wrap (modulo) the animation length, which didn't I found to be very unclear and hard to track the frame.

I removed the modulo wrapping guard - now if you render frame 51 on a 50 frame animation, you will see nothing because no layers apply to frame 51.

In general I do see there is value in render_t or something with modulo to expose the previous behavior, though.

Thoughts?

Comment (dfrg) [reply src/runtime/render.rs:83]: Completely agree that we eventually want more sophisticated control over time to frame mapping supporting different looping behaviors. I think it makes sense to punt that to a future PR where we can focus on the details and get it right. For this PR, I’d just prefer not to have two methods that do the same thing so let’s remove one of them. Which name to keep is up to you.

File: src/parser/schema/animated_properties/keyframe.rs
Comment (simbleau) [reply src/parser/schema/animated_properties/keyframe.rs]: Fixed

File: src/import/converters.rs
Comment (simbleau) [reply src/import/converters.rs]: Great suggestion, applied.

File: src/runtime/model/value.rs
Comment (simbleau) [reply src/runtime/model/value.rs]: Done

File: src/runtime/model/value.rs
Comment (simbleau) [reply src/runtime/model/value.rs:90]: Can you link to the Skottie implementation for reference?

File: src/runtime/model/fixed.rs
Comment (simbleau) [reply src/runtime/model/fixed.rs]: Filed this under peniko https://github.com/linebender/peniko/issues/27
And added as a followup task in https://github.com/linebender/velato/issues/14

File: src/runtime/render.rs
Comment (simbleau) [reply src/runtime/render.rs]: I consolidated these, choosing to keep frame instead of time.

File: src/runtime/model/value.rs
Comment (simbleau) [reply src/runtime/model/value.rs:165]: Added as a followup: https://github.com/linebender/velato/issues/14

File: src/lib.rs
Comment (simbleau) [reply src/lib.rs]: Fixed

File: src/lib.rs
Comment (simbleau) [reply src/lib.rs]: Fixed

File: src/import/util.rs
Comment (simbleau) [reply src/import/util.rs]: Fixed, wow this took a while 🥵

File: src/runtime/model/animated.rs
Comment (simbleau) [reply src/runtime/model/animated.rs]: Added as a followup in #14

File: src/runtime/model/animated.rs
Comment (simbleau) [reply src/runtime/model/animated.rs]: I think it was originally wrong, btw.
See this commit that "fixed" the skew logic: https://github.com/loopystudios/vellottie/commit/aaab5c52bf3862cf1b6348813ab390c458b7d29e

I just used `Affine::skew` at least, that works.

Was there a better implementation you had in mind?

I also have no idea what that SKEW_LIMIT you wrote in does. It seems when the skew is over that limit I get bad results, too. Here's an example file with 120 skew that looks *too skewed* in velato:

[lottie(2).json](https://github.com/linebender/velato/files/14711921/lottie.2.json)

File: src/runtime/model/value.rs
Comment (simbleau) [reply src/runtime/model/value.rs:90]: Either way, I added this as a task to #14 - Will resolve this one.

File: examples/with_winit/src/hot_reload.rs
Comment (simbleau) [reply examples/with_winit/src/hot_reload.rs:1]: This is tracked as a followup task

File: src/runtime/model/animated.rs
Comment (dfrg) [reply src/runtime/model/animated.rs]: Just using `Affine::skew` was all I intended here. I took the skew limit from Skottie but it might be wrong. It’s worth looking at other implementations but no need to block this PR on that.

File: src/import/util.rs
Comment (dfrg) [reply src/import/util.rs]: I thought this might be a big change but it seemed like a major simplification that would save both time and space. Thanks for addressing it.

File: src/runtime/model/value.rs
Comment (dfrg) [reply src/runtime/model/value.rs:90]: I’ll add a comment with the link to that issue.

**Review by dfrg — APPROVED** [review]
There’s a huge amount of great work here and I’m very happy to see it land upstream. Thanks!

**Pull request merged**
